### PR TITLE
allow 1 non-null value in interpolate_na with method="nearest"

### LINKED ISSUE #5994: interpolate_na, x and y arrays must have at least 2 entries, warning instead of raise ?
Hi,

With interpolate_na, the following behavior looks like it is non-ideal : 
```python
import xarray as xr
import numpy as np

# Example of data showing interpolate_na bug
x = np.arange(5)
y = np.arange(5)

data = np.array([[np.nan, np.nan, np.nan, np.nan, np.nan], 
                 [10, np.nan, np.nan, np.nan, np.nan], 
                 [10, 20, np.nan, np.nan, np.nan], 
                 [5, 3, np.nan, np.nan, 5], 
                 [5, 5, 5, 5, 5]])
print(data[1][1])

ds = xr.Dataset(
    data_vars=dict(
        data=(["x", "y"], data),
    ),
    coords=dict(
        x=(["x"], x),
        y=(["y"], y),
    ),
)

# interpolate_na will raise an exception
ds = ds.interpolate_na(dim="y", method="nearest")

# If we replace the np.nan in [1][1] with a valid value, the exception is gone
ds["data"][1][1] = 15

# Data now looks like
# data = np.array([[np.nan, np.nan, np.nan, np.nan, np.nan], 
#                 [10, 15, np.nan, np.nan, np.nan], 
#                 [10, 20, np.nan, np.nan, np.nan], 
#                 [5, 3, np.nan, np.nan, 5], 
#                 [5, 5, 5, 5, 5]])

# working interpolate_na
ds = ds.interpolate_na(dim="y", method="nearest")
```

**What can be expected**

In the case when interpolate_na raise an exception, we could instead expect either a warning or that the problematic row is ignored (like full nan rows are ignored)

See the below code which is part of xarray, which is called when using interpolate_na with method="nearest" (at least) and which calls scipy interp1d.

in xarray/core/missing.py
```python
def func_interpolate_na(interpolator, y, x, **kwargs):
    """helper function to apply interpolation along 1 dimension"""
    # reversed arguments are so that attrs are preserved from da, not index
    # it would be nice if this wasn't necessary, works around:
    # "ValueError: assignment destination is read-only" in assignment below
    out = y.copy()

    nans = pd.isnull(y)
    nonans = ~nans

    # fast track for no-nans and all-nans cases
    n_nans = nans.sum()

    if n_nans < 0 or n_nans >= len(y):
        return y

    f = interpolator(x[nonans], y[nonans], **kwargs)
    out[nans] = f(x[nans])
    return out
```

This is the
```python
    if n_nans < 0 or n_nans == len(y):
        return y
```
part that is interesting for me. What if the condition is replaced with `if n_nans < 0 or n_nans >= len(y) -1:`  ?

**Environment**:

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.10 (default, Jun  4 2021, 14:48:32) 
[GCC 7.5.0]
python-bits: 64
OS: Linux
OS-release: 4.15.0-142-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: fr_FR.UTF-8
LOCALE: ('fr_FR', 'UTF-8')
libhdf5: 1.12.0
libnetcdf: 4.7.4

xarray: 0.19.0
pandas: 1.2.5
numpy: 1.20.0
scipy: 1.7.1
netCDF4: 1.5.7
pydap: None
h5netcdf: None
h5py: 2.10.0
Nio: None
zarr: None
cftime: 1.5.1
nc_time_axis: None
PseudoNetCDF: None
rasterio: 1.1.0
cfgrib: None
iris: None
bottleneck: 1.3.2
dask: 2021.09.1
distributed: 2021.09.1
matplotlib: 3.1.3
cartopy: 0.18.0
seaborn: None
numbagg: None
pint: None
setuptools: 52.0.0.post20210125
pip: 21.1.3
conda: 4.10.3
pytest: None
IPython: 7.28.0
sphinx: None
</details>

### PR DESCRIPTION
**Pull request by delgadom**
- [x] Closes #5994 
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

@Skealz's great issue (#5994) highlighted that DataArrays with a single non-null value raise an error if `DataArray.interpolate_na` is called with `method='nearest'`. Note that this does not occur for the default (`method='linear'`).

In this PR:
* Uses the "fast track" to quickly return *any* 1-D slices with a single non-null value in addition to the previous all-nan or no-nan fasttrack
* Explicitly check this behavior for all 1-D interpolation methods using parametrized tests
* consolidate other fast-track tests (`interpolate_allnans` and `interpolate_nonans`) into a single parameterized test

For discussion:
* This patch may break existing workflows if they rely on the current behavior, which errors for slices with only a single value

### CODE REVIEW

**Comment by max-sixty** [comment]
Another good one — thanks @delgadom . 

Does anyone know this code better than me and could take a look? Otherwise let's merge soon.

File: xarray/tests/test_missing.py [hunk 1]
```diff
@@ -255,19 +255,30 @@ def test_interpolate():
...
+        pytest.param(method, vals, id=f"{desc}:{method}")
+        for method in [
+            "linear",
+            "nearest",
+            "zero",
+            "slinear",
+            "quadratic",
+            "cubic",
+            "polynomial",
+        ]
```
Comment (Illviljan) [inline xarray/tests/test_missing.py:271]: Would be nice if this wasn't hardcoded. But they seem hidden inside a function currently:
https://github.com/pydata/xarray/blob/5b322c9ea18f560e35857edcb78efe4e4f323551/xarray/core/missing.py#L471-L486

**Review by Illviljan — APPROVED** [review]
This makes sense to me. I only have one small thought about this and it's no blocker.

File: xarray/tests/test_missing.py (hunk 1 above)
Comment (delgadom) [reply xarray/tests/test_missing.py:271]: thanks @Illviljan ! yeah I considered trying to import these but opted not to modify xarray/core. but if you think creating a global variable or something is preferable I'm happy to add that to the pr!

Comment (Illviljan) [reply xarray/tests/test_missing.py:271]: I think we'll skip this for now. We'll add it when it causes problems.

**Comment by mathause** [comment]
Thanks @delgadom!

**Pull request merged**